File: QID-1227-MVAportfol_IBM_Ford/MVAportfol_IBM_Ford.r

# clear all variables
rm(list = ls(all = TRUE))
graphics.off()

# load data
ibm  = read.csv("ibm.csv")
ford = read.csv("ford.csv")

# compute returns for IBM
y1 = ibm[, 5]
a  = 0
i  = 1
while (i <= 120) {
    i    = i + 1
    a[i] = (y1[i] - y1[i - 1])/y1[i]
}
x1 = a[2:121]

# compute returns for Forward Industries
y2 = ford[, 5]
f  = 0
i  = 1
while (i <= 120) {
    i    = i + 1
    f[i] = (y2[i] - y2[i - 1])/y2[i]
}
x2 = f[2:121]

x   = cbind(x1, x2)                         # MVAportfolio of assets
s1  = solve(cov(x))                         # inverse of empirical variance
one = c(1, 1)                               # vector of ones
c2  = (s1 %*% one)/rep(t(one) %*% s1 %*% one, length(s1 %*% one)) 	# c2 weight
c1  = one/rep(sum(one), length(one))        # c1 weight
q1  = x %*% c1                              # Optimal MVAportfol_IBM_Ford returns 
q2  = x %*% c2                              # Nonoptimal MVAport_IBM_Ford returns 

t   = c(1:120)
d1  = cbind(t, q1)
d2  = cbind(t, q2)

# plot
par(mfrow = c(2, 1))
par(oma = c(0, 0, 0, 8))
plot(d1, type = "l", col = "blue", ylab = "Y", xlab = "X", main = "Equally Weighted Portfolio", 
    ylim = c(-0.6, 0.6), cex.lab = 1.4, cex.axis = 1.4, cex.main = 1.4)
mtext("Weights", side = 4, line = 4, at = 1.2, las = 2, font = 1)
mtext(toString(c(sprintf("%.3f", c1[1]), "IBM")), side = 4, line = 4, at = 0.6, las = 2)
mtext(toString(c(sprintf("%.3f", c1[2]), "Ford")), side = 4, line = 4, at = 0.45, 
    las = 2)
plot(d2, type = "l", col = "blue", ylab = "Y", xlab = "X", main = "Optimal Weighted Portfolio", 
    ylim = c(-0.6, 0.6), cex.lab = 1.4, cex.axis = 1.4, cex.main = 1.4)
mtext("Weights", side = 4, line = 4, at = 1.2, las = 2)
mtext(toString(c(sprintf("%.3f", c2[1]), "IBM")), side = 4, line = 4, at = 0.6, las = 2)
mtext(toString(c(sprintf("%.3f", c2[2]), "Ford")), side = 4, line = 4, at = 0.45, 
    las = 2) 

File: QID-1227-MVAportfol_IBM_Ford/MVAportfol_IBM_Ford.py
# Works on pandas 2.0.0, numpy 1.24.2, matplotlib 3.7.1
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Load data
ibm = pd.read_csv("ibm.csv")
ford = pd.read_csv("ford.csv")

# Function to compute returns
def compute_returns(df):
    y = df.iloc[:, 4]
    a = [0]
    for i in range(1, 121):  
        a.append((y[i] - y[i - 1]) / y[i])
    return a[1:]

# Compute returns for IBM and Forward Industries
x1 = compute_returns(ibm)
x2 = compute_returns(ford)

# MVAportfolio of assets
x = np.column_stack((x1, x2))

# Inverse of empirical variance
s1 = np.linalg.inv(np.cov(x.T))

# Vector of ones
one = np.ones(2)

# c2 weight
c2 = np.dot(s1, one) / np.dot(one.T, np.dot(s1, one))

# c1 weight
c1 = one / np.sum(one)

# Optimal MVAportfol_IBM_Ford returns
q1 = np.dot(x, c1)

# Nonoptimal MVAport_IBM_Ford returns
q2 = np.dot(x, c2)

# Time Series
t = np.arange(1, 121)
ts1 = pd.DataFrame({'Time': t, 'Returns': q1})
ts2 = pd.DataFrame({'Time': t, 'Returns': q2})


plt.figure(figsize=(8, 8))

# Equally Weighted Portfolio
plt.subplot(2, 1, 1)
plt.plot(ts1['Time'], ts1['Returns'], color='blue', linewidth=0.7)
plt.title('Equally Weighted Portfolio', fontweight='bold')
plt.xlabel('X')
plt.ylabel('Y')
plt.ylim(-0.6, 0.6)
plt.yticks([-0.5, 0, 0.5])
plt.text(ts1['Time'].iloc[-1] + 15, 0, f"Weights\n\n\n\n{c1[0]:.3f}, IBM\n{c1[1]:.3f}, Ford", 
         verticalalignment='center')

# Optimal Weighted Portfolio
plt.subplot(2, 1, 2)
plt.plot(ts2['Time'], ts2['Returns'], color='blue', linewidth=0.7)
plt.title('Optimal Weighted Portfolio', fontweight='bold')
plt.xlabel('X')
plt.ylabel('Y')
plt.ylim(-0.6, 0.6)
plt.yticks([-0.5, 0, 0.5])
plt.text(ts2['Time'].iloc[-1] + 15, 0, f"Weights\n\n\n\n{c2[0]:.3f}, IBM\n{c2[1]:.3f}, Ford", 
         verticalalignment='center')

plt.tight_layout()
plt.show()